litgpt/adapter.py

```diff
@@ -94,11 +97,21 @@ class Block(BaseBlock):
     def __init__(self, config: Config, block_idx: int) -> None:
         # Skip the parent class __init__ altogether and replace it to avoid useless allocations
         nn.Module.__init__(self)
+        if not config.parallel_residual and config.shared_attention_norm:
+            raise NotImplementedError(
+                "No checkpoint amongst the ones we support uses this configuration"
+                " (non-parallel residual and shared attention norm)."
+            )
         self.norm_1 = config.norm_class(config.n_embd, eps=config.norm_eps)
         self.attn = CausalSelfAttention(config, block_idx)
-        if not config.shared_attention_norm:
-            self.norm_2 = config.norm_class(config.n_embd, eps=config.norm_eps)
+        self.post_attention_norm = (
+            config.norm_class(config.n_embd, eps=config.norm_eps) if config.post_attention_norm else nn.Identity()
```

Comment:
I like the way this is handled. Not a priority since we don't have a model without pre-LN, but a similar setting for pre-LN could be nice for consistency. (Doesn't have to be done in this PR)

Reply:
Yes, that's what I also have been thinking about.
A nice set of:
1. pre attention norm
2. post attention norm
3. pre MLP norm
4. post MLP norm

Looks and reads much easier than the current norm_1 and norm_2.

But it's definitely a topic for a separate PR.
This requires a lot of changes in the conversion scripts + we need to handle legacy checkpoints.
I'll create an issue with a list of what needs to be done after this PR is merged.